### PR TITLE
Use direct Places query instead of autocomplete service.

### LINKED ISSUE #133: Update search service and autocomplete service handling in FF 43+
In current Nightly (45), the best search suggestion is included in the autocomplete search results as a `moz-action:searchengine` subtype, and we don't seem to have any others available (at least, not using our current code which queries that service):

screenshot 2015-10-30 14 45 07

Contrast this with the results returned in release channel (FF 41):

screenshot 2015-10-30 14 47 46

This is going to start out as a research bug, and might need to be split out into a few little implementation bugs. Questions for right now:
- What's the right way to get a complete list of search suggestions in Nightly?
- What's the first version of FF in which this stuff has landed? Is it in Dev edition?
- Is it possible in Nightly to ask the autocomplete service to exclude search suggestions from the results returned?

### LINKED ISSUE #139: Favicons: send favicon or globe image with each autocomplete result
We currently use a prototype favicon proxy service[1] to grab favicons. We don't want to keep using it.

Send over a good favicon or the standard FF globe image with each autocomplete result. We might need to figure out how to load the image into memory & serialize to JSON, so it can be sent over the postmessage barrier. Surely XPCOM has something to do this.

[1] https://github.com/nchapman/summarizer-server

### LINKED ISSUE #140: Include frecency value with places suggestions
When we send the history/bookmarks over, we should include the frecency value for that particular visit.

This will make it easier to blend results on the front-end.

### LINKED ISSUE #143: Places results sometimes omit the type
All of the relevant `type` fields should be set when we return results. We might want to do some de-duping on the addon side, if we can't figure out how to change the query so that we get the whole type on each result.

Just saw this when looking at some content changes, which is unfortunate.

Note that the `bookmark` type is not included in the second result:

```
{"type":"autocomplete-search-results","data":[
  {"url":"moz-action:switchtab,http://www.foo.com/","image":null,"title":"Foo.com",
    "type":"action bookmark","text":"foo"},
  {"url":"http://foo.com/","image":null,"title":"Foo.com",
    "type":"favicon","text":"foo"}
```

This might be fixed when we switch to using the Places DB directly, but let's keep this open until it's confirmed.

### PR DESCRIPTION
**Pull request by jaredhirsch**
Include frecency and favicon image (as base64 data) in responses.

Fixes #133, fixes #140, fixes #139, fixes #143.

(Still need to please the linter and double-check that the fixed bugs are really fixed.)

@chuckharmston give it as much r? as seems reasonable :-)

### CODE REVIEW

File: src/chrome/content/content.xml [hunk 1]
```diff
@@ -6,29 +6,49 @@
           xmlns:xbl="http://www.mozilla.org/xbl">
   <binding id="autocomplete-rich-result-popup-univ-search" extends="chrome://global/content/bindings/autocomplete.xml#autocomplete-rich-result-popup">
     <content ignorekeys="false" level="top" consumeoutsideclicks="false">
```
Comment (jaredhirsch) [inline src/chrome/content/content.xml:8]: Added some much-needed documentation to the craziest corner of our codebase :-)

File: src/lib/ui/Popup.js [hunk 2]
```diff
@@ -126,64 +128,30 @@ Popup.prototype = {
...
-      this._getSearchSuggestions().then((searchSuggestions) => {
-        this.app.broker.publish('popup::autocompleteSearchResults', autocompleteResults);
-
-        delete searchSuggestions.formHistoryResult;
-        this.app.broker.publish('popup::suggestedSearchResults',
-                                 searchSuggestions);
-      }, (err) => {
-        Cu.reportError(err);
-        this.app.broker.publish('popup::autocompleteSearchResults', autocompleteResults);
+    let placesResults;
```
Comment (jaredhirsch) [inline src/lib/ui/Popup.js]: This part of the diff looks pretty nuts, but mostly because of indentation changes.

File: src/lib/PlacesSearch.js [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+    const openUrls = [];
+    let browserState;
+
+    // Get a list of all open tabs from nsSessionStore, then look for the specified
+    // url in the list.
+    // If anything goes wrong with parsing the browser state JSON, just give up
+    // and return a falsy value.
+    try {
+      browserState = JSON.parse(SessionStore.getBrowserState());
+    } catch (ex) {}
```
Comment (pdehaan) [inline src/lib/PlacesSearch.js]: Travis CI [says](https://travis-ci.org/mozilla/universal-search-addon/builds/94030842#L496):

```
[21:17:25] 'eslint' errored after 540 ms
[21:17:25] ESLintError in plugin 'gulp-eslint'
Message:
    Empty block statement.
Details:
    fileName: /home/travis/build/mozilla/universal-search-addon/src/lib/PlacesSearch.js
    lineNumber: 179
```

Comment (jaredhirsch) [inline src/lib/PlacesSearch.js]: Yeah, pushed this before linting. At least that much should now be fixed

**Comment by jaredhirsch** [comment]
OK, seems to work for me, merging!

**Comment by chuckharmston** [comment]
Sorry I didn't get to this in time!

**Pull request merged**